docs/lazy-adr/adr-002-ipds-da-sampling.md

```diff
@@ -0,0 +1,276 @@
+// TODO: Should there be a constant = lower bound for #samples
+func ValidateAvailability(
+    ctx contex.Context,
+    dah *DataAvailabilityHeader,
+    numSamples int,
+    leafSucessCb func(namespacedleaf []byte),
+) error { /* ... */}
+
+// RetrieveBlock can be used to recover the block Data.
+// It will carry out a similar protocol as described for ValidateAvailability.
+// The key difference is that it will sample enough chunks until it can recover the
+// original data.
+func RetrieveBlock(ctx contex.Context, dah *DataAvailabilityHeader) (types.Data, error) {/* ... */}
+
+// PutLeaves takes the namespaced leaves from the extended data square and calls
+// nodes.DataSquareRowOrColumnRawInputParser of the ipld plugin.
+// The resulting ipld nodes are passed to a Batch calling AddMany:
+// https://github.com/ipfs/go-ipld-format/blob/d2e09424ddee0d7e696d01143318d32d0fb1ae63/batch.go#L29
+// Note, that this method could also return the row and column roots.
+// Tha caller is responsible for making sure that the leaves are sorted by namespace ID.
+// The data will be pinned by default.
+func PutLeaves(ctx contex.Context, namespacedLeaves [][]byte) error
+```
+
+As an alternative to the above `PutLeaves` method, the API could also just take in a block.
+Then it would also need to internally do the erasure coding.
+
+We now describe the lower-level library that will be used by above methods.
+Again we provide more details inline in the godoc comments directly.
+
+```go
+// GetLeafData takes in a Namespaced Merkle tree root transformed into a Cid
+// and the leaf index to retrieve.
+// Callers also need to pass in the total number of leaves of that tree.
+// Internally, this will be translated to a path and corresponds to
+// a ipfs dag get request, e.g. namespacedCID/0/1/0/0/1.
+// The retrieved data will be pinned by default.
+func GetLeafData(
+    ctx context.Context,
+    rootCid cid.Cid,
+    leafIndex uint32,
+    totalLeafs uint32, // this corresponds to the extended square width
+) ([]byte, error)
+```
+
+> TODO: add method that calls GetLeafData often enough until the original data can be reconstructed.
+
+`GetLeafData` can be used by above `ValidateAvailability` and `RetrieveBlock`.
+We do not define a corresponding put-function as above's `PutLeaves` already
+
+### A Note on IPFS/IPLD
+
+In IPFS all data is _content addressed_ which basically means the data is identified by its hash.
+Particularly, in the LazyLedger case, the root CID identifies the Namespaced Merkle tree including all its contents (inner and leaf nodes).
+This means that if a `GetLeafData` request succeeds, the retrieved leaf data is in fact the leaf data in the tree.
+We do not need to additionally verify Merkle proofs per leaf as this will essentially be done via IPFS on each layer while
+resolving and getting to the leaf data.
+
+> TODO: validate this assumption and link to code that shows how this is done internally
```

Comment:
Or do we want to explicitly verify proofs either way? To not rely on the fact that ipfs in combination with our plugin handles this correctly?

cc @musalbas @adlerjohn

Reply:
I think it's safer and more idiot-proof if `GetLeafData`only succeeds if the proof is valid. Anyway, I thought it would only succeed with IPFS is the proof is valid, with the custom hasher?

Reply:
> Anyway, I thought it would only succeed with IPFS is the proof is valid, with the custom hasher?

Yes, that is my understanding as well. For every retrieved leaf, the proof nodes should also be resolved and validated on its path down.